core/src/common.cpp

```diff
@@ -362,6 +364,264 @@ namespace irods::http
+
+		// Get the JWK the access token was signed with. This is optional.
+		// See RFC 7515 Section 4.1.4
+		try {
+			auto jwk{_jwks.get_jwk(decoded_token.get_key_id())};
+			add_alg_from_jwk(_verifier, jwk, _alg);
+		}
+		catch (const std::runtime_error& e) {
+			// We cannot pick out the specific key used...
+			// Find JWKs that match any of the following:
+			//   - 'alg' (optional) matches exactly
+			//   - 'kty' (required) matches 'alg' family [EC,RSA,oct]
+			//   - 'use' (required if sign & encrypt keys exist) to be 'sig'
+			//
+			// Additionally, symetric algoritms it /seems/ that
+			// "... the octects of the 'client_secret'..." are used for the algorithm...
+			const auto algorithm_family{_alg.substr(0,2)};
+			std::string search_string;
+
+			// RSA family
+			if (algorithm_family == "RS") {
+				search_string = "RSA";
+			}
+			// EC family
+			else if (algorithm_family == "ES") {
+				search_string = "EC";
+			}
+			// Symmetric algo
+			else if (algorithm_family == "HS") {
+				search_string = "oct";
+			}
+
+			// Go through entire key set
+			std::for_each(std::cbegin(_jwks), std::cend(_jwks), [&_verifier, &_alg, &search_string] (const auto& _jwk) -> void {
+				try {
+					if (auto key_type{_jwk.get_jwk_claim("kty").as_string()}; key_type == search_string) {
+						add_alg_from_jwk(_verifier, _jwk, _alg);
+					}
+				}
+				catch (const std::runtime_error& e) {
+					logging::error("{}: Invalid key [{}]", __func__, e.what());
+					return;
+				}
+			});
+		}
+
+		// Allow for chaining
+		return _verifier;
+	}
+	
+	/// Validates an OAuth 2.0 Access Token using
+	///
+	/// See RFC 9068 for more details on JWT for OAuth 2.0
+	/// See RFC 7515 for details on JSON Web Signature (JWS)
+	/// See RFC 7518 for details on JSON Web Algorithms (JWA)
+	///
+	/// \param[in] _jwt A std::string represeting the JWT to verify.
+	///
+	/// \returns The JWT payload if the token can be validated. Otherwise, an empty std::optional is returned
+	///
+	/// \todo Handle encrypted tokens. MUST reject unencrypted tokens if encryption was negotiated at registration.
+	/// \date 2024-10-16 jwt-cpp does not support encrypted tokens as of current.
+	auto validate_using_local_validation(std::string _jwt) -> std::optional<nlohmann::json> {
+		namespace logging = irods::http::log;
+
+		// Decode the JWT
+		auto decoded_token{jwt::decode<jwt::traits::nlohmann_json>(_jwt)};
+
+		// Parse the JWKs discoverd from the OpenID Provider
+		static auto jwks{jwt::parse_jwks<jwt::traits::nlohmann_json>(fetch_jwks_from_openid_provider())};
+
+		// Manually verify 'typ' matches what is specified in RFC 9068
+		// Validate 'typ' is 'at+jwt' or 'application/at+jwt', reject all others
+		try {
+			auto token_type{decoded_token.get_header_claim("typ").as_string()};
+			if (!(token_type == "at+jwk" || token_type == "application/at+jwk")) {
+				logging::error("{}: Access Token with [typ] of type [{}] is not supported.", __func__, token_type);
+				// return std::nullopt;
```

Comment:
please confirm or delete

Reply:
Commented for debugging. Uncommenting for later.

Reply:
Uncommented.